In [67]:
import pandas as pd 
import numpy as np 
import os 

print(os.getcwd())

/Users/eugenechua/Downloads/course_skill_proficiency_evaluator


In [66]:
# Reading in the course skill extractions.
course_skill_df = pd.read_excel('./data/final_course_skills_merged.xlsx')
print(course_skill_df.shape)
print(course_skill_df['coursereferencenumber'].nunique())
course_skill_df.head(5)

(916, 12)
794


,coursereferencenumber,coursetitle,coursedescription,skillTitle,skillSource,proficiencyLevel,proficiencyScore,confidenceScore,skillOccurrence,phrase,uniqueskilltitle,proficiencylevelofcourse
0,TGS-2019503800,Combi Oven Cooking for Chinese Cuisine,"In this hands-on workshop, participants will l...",Chinese Dry-Heat Dish Preparation and Cooking,tsc,Intermediate,0.8737,0.677110,1,full text,Chinese Dry-Heat Dish Preparation and Cooking,Basic
1,TGS-2019503804,Social Media for Food Marketing,Particpants will examine the use of social med...,Social Media Marketing,tsc,Basic,0.7964,0.731026,1,full text,Social Media Marketing,Basic
2,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Governance,tsc,Advanced,0.8033,0.890494,1,data governance,Data Governance,Intermediate
3,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8406,0.694061,1,full text,Data Strategy,Intermediate
4,TGS-2020001034,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8406,0.694061,1,full text,Data Strategy,Intermediate


In [4]:
course_skill_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 916 entries, 0 to 915
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   coursereferencenumber     916 non-null    object 
 1   coursetitle               916 non-null    object 
 2   coursedescription         916 non-null    object 
 3   skillTitle                916 non-null    object 
 4   skillSource               916 non-null    object 
 5   proficiencyLevel          916 non-null    object 
 6   proficiencyScore          871 non-null    float64
 7   confidenceScore           916 non-null    float64
 8   skillOccurrence           916 non-null    int64  
 9   phrase                    915 non-null    object 
 10  uniqueskilltitle          916 non-null    object 
 11  proficiencylevelofcourse  916 non-null    object 
dtypes: float64(2), int64(1), object(9)
memory usage: 86.0+ KB


In [5]:
course_skill_df.isnull().sum()

coursereferencenumber        0
coursetitle                  0
coursedescription            0
skillTitle                   0
skillSource                  0
proficiencyLevel             0
proficiencyScore            45
confidenceScore              0
skillOccurrence              0
phrase                       1
uniqueskilltitle             0
proficiencylevelofcourse     0
dtype: int64

In [6]:
# Reading in course skill with updated algo
course_skill_updated_algo_df = pd.read_excel('./data/final_course_skills_merged_modified_algo.xlsx')
print(course_skill_updated_algo_df.shape)
print(course_skill_updated_algo_df['coursereferencenumber'].nunique())
course_skill_updated_algo_df.head(5)

(916, 12)
794


,coursereferencenumber,coursetitle,coursedescription,skillTitle,skillSource,proficiencyLevel,proficiencyScore,confidenceScore,skillOccurrence,phrase,uniqueskilltitle,proficiencylevelofcourse
0,TGS-2019503800,Combi Oven Cooking for Chinese Cuisine,"In this hands-on workshop, participants will l...",Chinese Dry-Heat Dish Preparation and Cooking,tsc,Intermediate,0.8566,0.677110,1,full text,Chinese Dry-Heat Dish Preparation and Cooking,Basic
1,TGS-2019503804,Social Media for Food Marketing,Particpants will examine the use of social med...,Social Media Marketing,tsc,Advanced,0.8116,0.731026,1,full text,Social Media Marketing,Basic
2,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Governance,tsc,Advanced,0.8171,0.890494,1,data governance,Data Governance,Intermediate
3,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8295,0.694061,1,full text,Data Strategy,Intermediate
4,TGS-2020001034,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8295,0.694061,1,full text,Data Strategy,Intermediate


In [7]:
course_skill_updated_algo_df.isnull().sum()

coursereferencenumber        0
coursetitle                  0
coursedescription            0
skillTitle                   0
skillSource                  0
proficiencyLevel             0
proficiencyScore            45
confidenceScore              0
skillOccurrence              0
phrase                       1
uniqueskilltitle             0
proficiencylevelofcourse     0
dtype: int64

In [8]:
course_skill_updated_algo_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 916 entries, 0 to 915
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   coursereferencenumber     916 non-null    object 
 1   coursetitle               916 non-null    object 
 2   coursedescription         916 non-null    object 
 3   skillTitle                916 non-null    object 
 4   skillSource               916 non-null    object 
 5   proficiencyLevel          916 non-null    object 
 6   proficiencyScore          871 non-null    float64
 7   confidenceScore           916 non-null    float64
 8   skillOccurrence           916 non-null    int64  
 9   phrase                    915 non-null    object 
 10  uniqueskilltitle          916 non-null    object 
 11  proficiencylevelofcourse  916 non-null    object 
dtypes: float64(2), int64(1), object(9)
memory usage: 86.0+ KB


In [72]:
# Reading in the course skills with chatgpt proficiency labelling to see what it looks like
course_skill_gpt_labelled = pd.read_csv('./data/courses_skills_with_proficiency_labels.csv')
print(course_skill_gpt_labelled.shape)
print(course_skill_gpt_labelled['coursereferencenumber'].nunique())
course_skill_gpt_labelled.head(10)

(916, 16)
794


,coursereferencenumber,coursetitle,coursedescription,skillTitle,skillSource,proficiencyLevel,proficiencyScore,confidenceScore,skillOccurrence,phrase,uniqueskilltitle,proficiencylevelofcourse,gpt_proficiencyLevel,gpt_rationale,gpt_raw_response,gpt_error
0,TGS-2019503800,Combi Oven Cooking for Chinese Cuisine,"In this hands-on workshop, participants will l...",Chinese Dry-Heat Dish Preparation and Cooking,tsc,Intermediate,0.8737,0.677110,1,full text,Chinese Dry-Heat Dish Preparation and Cooking,Basic,Basic,The course description states that participant...,"{\n ""skillTitle"": ""Chinese Dry-Heat Dish Prep...",NaN
1,TGS-2019503804,Social Media for Food Marketing,Particpants will examine the use of social med...,Social Media Marketing,tsc,Basic,0.7964,0.731026,1,full text,Social Media Marketing,Basic,Intermediate,The course description uses the word 'examine'...,"{\n ""skillTitle"": ""Social Media Marketing"",\n...",NaN
2,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Governance,tsc,Advanced,0.8033,0.890494,1,data governance,Data Governance,Intermediate,Basic,The course description focuses on identifying ...,"{\n ""skillTitle"": ""Data Governance"",\n ""prof...",NaN
3,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8406,0.694061,1,full text,Data Strategy,Intermediate,Basic,The course description focuses on identifying ...,"{\n ""skillTitle"": ""Data Strategy"",\n ""profic...",NaN
4,TGS-2020001034,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8406,0.694061,1,full text,Data Strategy,Intermediate,Basic,The course description focuses on identifying ...,"{\n ""skillTitle"": ""Data Strategy"",\n ""profic...",NaN
5,TGS-2020001034,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Governance,tsc,Advanced,0.8033,0.890494,1,data governance,Data Governance,Intermediate,Basic,The course description focuses on learning abo...,"{\n ""skillTitle"": ""Data Governance"",\n ""prof...",NaN
6,TGS-2020501272,Train The Trainer - Develop and Conduct Assess...,•\tFormulate assessment scheme\r\n•\tDevel...,Assessment Design and Implementation,tsc,Intermediate,0.7604,0.797698,2,assessment scheme Develop theory assessment sp...,Assessment Design and Implementation,Intermediate,Advanced,The course requires learners to 'formulate ass...,"{\n ""skillTitle"": ""Assessment Design and Impl...",NaN
7,TGS-2020502315,Software Development Fundamentals,This course aims to introduce non-computing st...,Applications Development,tsc,Intermediate,0.5980,0.609969,1,full text,Applications Development,Intermediate,Basic,The course introduces foundational concepts an...,"{\n ""skillTitle"": ""Applications Development"",...",NaN
8,TGS-2020502315,Software Development Fundamentals,This course aims to introduce non-computing st...,Software Testing,tsc,Advanced,0.6417,0.651658,1,testing,Software Testing,Intermediate,Basic,The course description states students will be...,"{\n ""skillTitle"": ""Software Testing"",\n ""pro...",NaN
9,TGS-2020502315,Software Development Fundamentals,This course aims to introduce non-computing st...,Agile Software Development,tsc,Basic,0.6478,0.608859,3,software development practices,Agile Software Development,Intermediate,Basic,The course description states that students wi...,"{\n ""skillTitle"": ""Agile Software Development...",NaN


In [73]:
print(course_skill_gpt_labelled.isnull().sum())

coursereferencenumber         0
coursetitle                   0
coursedescription             0
skillTitle                    0
skillSource                   0
proficiencyLevel              0
proficiencyScore             45
confidenceScore               0
skillOccurrence               0
phrase                        0
uniqueskilltitle              0
proficiencylevelofcourse      0
gpt_proficiencyLevel          0
gpt_rationale                 0
gpt_raw_response              0
gpt_error                   916
dtype: int64


In [46]:
print(course_skill_gpt_labelled['gpt_proficiencyLevel'].value_counts())
print(course_skill_gpt_labelled['gpt_proficiencyLevel'].value_counts(normalize=True))

gpt_proficiencyLevel
Intermediate    418
Basic           364
Advanced        134
Name: count, dtype: int64
gpt_proficiencyLevel
Intermediate    0.456332
Basic           0.397380
Advanced        0.146288
Name: proportion, dtype: float64


In [47]:
print(course_skill_gpt_labelled['proficiencyLevel'].value_counts())
print(course_skill_gpt_labelled['proficiencyLevel'].value_counts(normalize=True))

proficiencyLevel
Intermediate    424
Advanced        302
Basic           190
Name: count, dtype: int64
proficiencyLevel
Intermediate    0.462882
Advanced        0.329694
Basic           0.207424
Name: proportion, dtype: float64


### Performing model evaluation now on both baseline and slightly modified models

In [48]:
# Doing validation for the baseline algo (non modified!)
from sklearn.metrics import f1_score, precision_score, classification_report

y_pred = np.array(course_skill_gpt_labelled['proficiencyLevel'])
y_true = np.array(course_skill_gpt_labelled['gpt_proficiencyLevel'])

# Performing encoding of labels!
label_map = {'Basic': 0, 'Intermediate': 1, 'Advanced': 2}
y_pred = np.array([label_map[l] for l in y_pred])
y_true = np.array([label_map[l] for l in y_true])

print(y_pred[:20])
print(y_true[:20])


[1 0 2 2 2 2 1 1 2 0 0 2 2 2 2 2 2 0 0 0]
[0 1 0 0 0 0 2 0 0 0 0 2 1 2 2 2 2 1 1 1]


In [49]:
f1 = f1_score(y_true, y_pred, average='weighted', zero_division=0)
precision = precision_score(y_true, y_pred, average='weighted', zero_division=0)

print(f'weighted f1 score is {f1}')
print(f'weighted precision score is {precision}')

weighted f1 score is 0.35967325273771955
weighted precision score is 0.40333682398978576


In [50]:
y_pred = np.array(course_skill_gpt_labelled['proficiencyLevel'])
y_true = np.array(course_skill_gpt_labelled['gpt_proficiencyLevel'])

print(classification_report(y_true, y_pred, target_names=['Basic', 'Intermediate', 'Advanced'], labels=['Basic', 'Intermediate', 'Advanced']))


              precision    recall  f1-score   support

       Basic       0.43      0.22      0.29       364
Intermediate       0.47      0.47      0.47       418
    Advanced       0.14      0.32      0.20       134

    accuracy                           0.35       916
   macro avg       0.35      0.34      0.32       916
weighted avg       0.40      0.35      0.36       916



In [74]:
# Reading in the course skills with chatgpt proficiency labelling (modified algo) to see what it looks like
course_skill_gpt_labelled = pd.read_csv('./data/courses_skills_with_proficiency_labels_modified_algo.csv')
print(course_skill_gpt_labelled.shape)
print(course_skill_gpt_labelled['coursereferencenumber'].nunique())
course_skill_gpt_labelled.head(10)

(916, 16)
794


,coursereferencenumber,coursetitle,coursedescription,skillTitle,skillSource,proficiencyLevel,proficiencyScore,confidenceScore,skillOccurrence,phrase,uniqueskilltitle,proficiencylevelofcourse,gpt_proficiencyLevel,gpt_rationale,gpt_raw_response,gpt_error
0,TGS-2019503800,Combi Oven Cooking for Chinese Cuisine,"In this hands-on workshop, participants will l...",Chinese Dry-Heat Dish Preparation and Cooking,tsc,Intermediate,0.8566,0.677110,1,full text,Chinese Dry-Heat Dish Preparation and Cooking,Basic,Basic,The course description states participants wil...,"{\n ""skillTitle"": ""Chinese Dry-Heat Dish Prep...",NaN
1,TGS-2019503804,Social Media for Food Marketing,Particpants will examine the use of social med...,Social Media Marketing,tsc,Advanced,0.8116,0.731026,1,full text,Social Media Marketing,Basic,Intermediate,The course description states that participant...,"{\n ""skillTitle"": ""Social Media Marketing"",\n...",NaN
2,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Governance,tsc,Advanced,0.8171,0.890494,1,data governance,Data Governance,Intermediate,Basic,The course description focuses on identifying ...,"{\n ""skillTitle"": ""Data Governance"",\n ""prof...",NaN
3,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8295,0.694061,1,full text,Data Strategy,Intermediate,Basic,The course description focuses on identifying ...,"{\n ""skillTitle"": ""Data Strategy"",\n ""profic...",NaN
4,TGS-2020001034,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8295,0.694061,1,full text,Data Strategy,Intermediate,Basic,The course description focuses on identifying ...,"{\n ""skillTitle"": ""Data Strategy"",\n ""profic...",NaN
5,TGS-2020001034,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Governance,tsc,Advanced,0.8171,0.890494,1,data governance,Data Governance,Intermediate,Basic,The course description focuses on learning abo...,"{\n ""skillTitle"": ""Data Governance"",\n ""prof...",NaN
6,TGS-2020501272,Train The Trainer - Develop and Conduct Assess...,•\tFormulate assessment scheme\r\n•\tDevel...,Assessment Design and Implementation,tsc,Advanced,0.7710,0.797698,2,assessment scheme Develop theory assessment sp...,Assessment Design and Implementation,Intermediate,Advanced,The course requires learners to 'formulate ass...,"{\n ""skillTitle"": ""Assessment Design and Impl...",NaN
7,TGS-2020502315,Software Development Fundamentals,This course aims to introduce non-computing st...,Agile Software Development,tsc,Intermediate,0.6915,0.608859,3,software development practices,Agile Software Development,Intermediate,Basic,The course description states that students wi...,"{\n ""skillTitle"": ""Agile Software Development...",NaN
8,TGS-2020502315,Software Development Fundamentals,This course aims to introduce non-computing st...,Software Testing,tsc,Intermediate,0.7392,0.651658,1,testing,Software Testing,Intermediate,Basic,The course description states students will be...,"{\n ""skillTitle"": ""Software Testing"",\n ""pro...",NaN
9,TGS-2020502315,Software Development Fundamentals,This course aims to introduce non-computing st...,Software Design,tsc,Basic,0.7077,0.711091,3,software development,Software Design,Intermediate,Basic,"The course introduces principles and concepts,...","{\n ""skillTitle"": ""Software Design"",\n ""prof...",NaN


In [75]:
print(course_skill_gpt_labelled.isnull().sum())

coursereferencenumber         0
coursetitle                   0
coursedescription             0
skillTitle                    0
skillSource                   0
proficiencyLevel              0
proficiencyScore             45
confidenceScore               0
skillOccurrence               0
phrase                        0
uniqueskilltitle              0
proficiencylevelofcourse      0
gpt_proficiencyLevel          0
gpt_rationale                 0
gpt_raw_response              0
gpt_error                   916
dtype: int64


In [57]:
print(course_skill_gpt_labelled['proficiencyLevel'].value_counts())
print(course_skill_gpt_labelled['proficiencyLevel'].value_counts(normalize=True))

proficiencyLevel
Intermediate    452
Advanced        290
Basic           174
Name: count, dtype: int64
proficiencyLevel
Intermediate    0.493450
Advanced        0.316594
Basic           0.189956
Name: proportion, dtype: float64


In [52]:
print(course_skill_gpt_labelled['gpt_proficiencyLevel'].value_counts())
print(course_skill_gpt_labelled['gpt_proficiencyLevel'].value_counts(normalize=True))

gpt_proficiencyLevel
Intermediate    418
Basic           365
Advanced        133
Name: count, dtype: int64
gpt_proficiencyLevel
Intermediate    0.456332
Basic           0.398472
Advanced        0.145197
Name: proportion, dtype: float64


In [53]:
# Doing validation for the MODIFIED baseline algo!)
from sklearn.metrics import f1_score, precision_score, classification_report

y_pred = np.array(course_skill_gpt_labelled['proficiencyLevel'])
y_true = np.array(course_skill_gpt_labelled['gpt_proficiencyLevel'])

# Performing encoding of labels!
label_map = {'Basic': 0, 'Intermediate': 1, 'Advanced': 2}
y_pred = np.array([label_map[l] for l in y_pred])
y_true = np.array([label_map[l] for l in y_true])

print(y_pred[:20])
print(y_true[:20])

[1 2 2 2 2 2 2 1 1 0 1 2 2 2 1 2 2 0 0 0]
[0 1 0 0 0 0 2 0 0 0 0 2 1 2 2 2 2 1 1 1]


In [54]:
f1 = f1_score(y_true, y_pred, average='weighted', zero_division=0)
precision = precision_score(y_true, y_pred, average='weighted', zero_division=0)

print(f'weighted f1 score is {f1}')
print(f'weighted precision score is {precision}')

weighted f1 score is 0.36855332060525553
weighted precision score is 0.4221563639481028


In [55]:
y_pred = np.array(course_skill_gpt_labelled['proficiencyLevel'])
y_true = np.array(course_skill_gpt_labelled['gpt_proficiencyLevel'])

print(classification_report(y_true, y_pred, target_names=['Basic', 'Intermediate', 'Advanced'], labels=['Basic', 'Intermediate', 'Advanced']))


              precision    recall  f1-score   support

       Basic       0.49      0.23      0.32       365
Intermediate       0.46      0.49      0.47       418
    Advanced       0.13      0.29      0.18       133

    accuracy                           0.36       916
   macro avg       0.36      0.34      0.32       916
weighted avg       0.42      0.36      0.37       916



Alright the performances of the 2 baseline model variants is quite abit underwhelming ~ 40% - 42% precision is not quite acceptable..In the meantime, I will try testing the SEA with PL extractor on job postings descriptions, before moving on to think about tweaking the model architecture.

The fundamental problem is that course description are quite often badly written. There are numerous instances where the course title states its advanced, but the description are in bullet points and very basic, hence making it impossibly hard for the PL extractor to even extract anything relevant..

### Job Skill Proficiency Labelling evaluation

In [192]:
job_skill_gpt_evaluation = pd.read_csv('./data/jobpostings_skills_with_proficiency_labels.csv')
print(job_skill_gpt_evaluation.shape)
print(job_skill_gpt_evaluation.isnull().sum())
job_skill_gpt_evaluation.head(5)

(7379, 14)
job_id                     0
jobtitle                   0
jobdescription             0
skillTitle                 0
skillSource                0
proficiencyLevel         165
proficiencyScore         498
confidenceScore            0
skillOccurrence            0
phrase                     0
gpt_proficiencyLevel    1185
gpt_rationale           1185
gpt_raw_response        1185
gpt_error               7173
dtype: int64


,job_id,jobtitle,jobdescription,skillTitle,skillSource,proficiencyLevel,proficiencyScore,confidenceScore,skillOccurrence,phrase,gpt_proficiencyLevel,gpt_rationale,gpt_raw_response,gpt_error
0,b59e719539de5ee2cbadd138a365a73714e3dbb2d0db6f...,"Engineer, Unix/Linux System - Up to $7100, Eas...",Mandatory Skills: Red Hat Enterprise Linux (RH...,System Architecture Design,tsc,Intermediate,NaN,0.695898,1,system architecture design,Intermediate,The job requires participation in system archi...,"{\n ""skillTitle"": ""System Architecture Design...",NaN
1,b59e719539de5ee2cbadd138a365a73714e3dbb2d0db6f...,"Engineer, Unix/Linux System - Up to $7100, Eas...",Mandatory Skills: Red Hat Enterprise Linux (RH...,Security Management,tsc,Basic,0.6981,0.608756,1,Implement security measures,Intermediate,The job requires implementing security measure...,"{\n ""skillTitle"": ""Security Management"",\n ""...",NaN
2,b59e719539de5ee2cbadd138a365a73714e3dbb2d0db6f...,"Engineer, Unix/Linux System - Up to $7100, Eas...",Mandatory Skills: Red Hat Enterprise Linux (RH...,Documentation and Administration,tsc,Intermediate,0.5850,0.604023,1,Write and maintain system administration docum...,Intermediate,The job requires writing and maintaining syste...,"{\n ""skillTitle"": ""Documentation and Administ...",NaN
3,f6d2a09934e94b899e4caffb7fd3db3775f54eed77fa38...,IT Business Development Manager #IFL,His/her mission is to demonstrate to potential...,Sales and Business Development,tsc,Intermediate,0.7886,0.706363,1,support sales development,Advanced,The job requires seeking out new opportunities...,"{\n ""skillTitle"": ""Sales and Business Develop...",NaN
4,f6d2a09934e94b899e4caffb7fd3db3775f54eed77fa38...,IT Business Development Manager #IFL,His/her mission is to demonstrate to potential...,Sales Management,tsc,Advanced,0.6913,0.639902,1,Develop and execute sales pipeline,Advanced,The job requires developing and executing sale...,"{\n ""skillTitle"": ""Sales Management"",\n ""pro...",NaN


The thing is because i was freaking dumb enough to not do multithreading and ran into alot of time out issues...so I had to manually filter out the remaining rows that were not processed by chatgpt.

In [193]:
remaining_rows_df = job_skill_gpt_evaluation[job_skill_gpt_evaluation['gpt_proficiencyLevel'].isna() & job_skill_gpt_evaluation['gpt_error'].isna()]
print(remaining_rows_df.shape)
remaining_rows_df.head(5)

(979, 14)


,job_id,jobtitle,jobdescription,skillTitle,skillSource,proficiencyLevel,proficiencyScore,confidenceScore,skillOccurrence,phrase,gpt_proficiencyLevel,gpt_rationale,gpt_raw_response,gpt_error
6400,158086cf40d8e65a42e90ec2a5ff90814dea936a24ccf9...,Senior-Engineer - Global Engineering Labs,Tanggungjawab You are responsible for new equi...,Mechanical Maintenance Management,tsc,Intermediate,0.7792,0.680266,1,"Oversees the installation, modification, upgra...",NaN,NaN,NaN,NaN
6401,158086cf40d8e65a42e90ec2a5ff90814dea936a24ccf9...,Senior-Engineer - Global Engineering Labs,Tanggungjawab You are responsible for new equi...,Equipment and Systems Repair,tsc,Advanced,0.8586,0.674246,2,spare parts planning,NaN,NaN,NaN,NaN
6402,158086cf40d8e65a42e90ec2a5ff90814dea936a24ccf9...,Senior-Engineer - Global Engineering Labs,Tanggungjawab You are responsible for new equi...,Equipment Maintenance,tsc,Basic,0.8498,0.670282,1,equipment usage,NaN,NaN,NaN,NaN
6403,158086cf40d8e65a42e90ec2a5ff90814dea936a24ccf9...,Senior-Engineer - Global Engineering Labs,Tanggungjawab You are responsible for new equi...,Preventive Maintenance Management,tsc,Advanced,0.8576,0.669205,1,Defines and writes preventative maintenance sc...,NaN,NaN,NaN,NaN
6404,158086cf40d8e65a42e90ec2a5ff90814dea936a24ccf9...,Senior-Engineer - Global Engineering Labs,Tanggungjawab You are responsible for new equi...,Corrective Maintenance Management,tsc,Advanced,0.8381,0.668778,1,full text,NaN,NaN,NaN,NaN


In [194]:
remaining_rows_df.isnull().sum()

job_id                    0
jobtitle                  0
jobdescription            0
skillTitle                0
skillSource               0
proficiencyLevel         23
proficiencyScore         54
confidenceScore           0
skillOccurrence           0
phrase                    0
gpt_proficiencyLevel    979
gpt_rationale           979
gpt_raw_response        979
gpt_error               979
dtype: int64

In [195]:
# Exporting the remaining rows
remaining_rows_df.to_csv('./data/extracted_jobs_skills_dataset_remaining.csv', index=False)

Now I will have to manually read in the processed 979 records that were previously not processed by chatgpt.

In [196]:
job_skill_gpt_evaluation_remaining = pd.read_csv('./data/jobpostings_skills_with_proficiency_labels_remaining.csv')
print(job_skill_gpt_evaluation_remaining.shape)
print(job_skill_gpt_evaluation_remaining.isnull().sum())
job_skill_gpt_evaluation_remaining.head(5)

(979, 14)
job_id                    0
jobtitle                  0
jobdescription            0
skillTitle                0
skillSource               0
proficiencyLevel         23
proficiencyScore         54
confidenceScore           0
skillOccurrence           0
phrase                    0
gpt_proficiencyLevel      0
gpt_rationale             0
gpt_raw_response          0
gpt_error               979
dtype: int64


,job_id,jobtitle,jobdescription,skillTitle,skillSource,proficiencyLevel,proficiencyScore,confidenceScore,skillOccurrence,phrase,gpt_proficiencyLevel,gpt_rationale,gpt_raw_response,gpt_error
0,158086cf40d8e65a42e90ec2a5ff90814dea936a24ccf9...,Senior-Engineer - Global Engineering Labs,Tanggungjawab You are responsible for new equi...,Mechanical Maintenance Management,tsc,Intermediate,0.7792,0.680266,1,"Oversees the installation, modification, upgra...",Advanced,The job requires establishing programs and sol...,"{\n ""skillTitle"": ""Mechanical Maintenance Man...",NaN
1,158086cf40d8e65a42e90ec2a5ff90814dea936a24ccf9...,Senior-Engineer - Global Engineering Labs,Tanggungjawab You are responsible for new equi...,Equipment and Systems Repair,tsc,Advanced,0.8586,0.674246,2,spare parts planning,Advanced,"The job requires overseeing installation, modi...","{\n ""skillTitle"": ""Equipment and Systems Repa...",NaN
2,158086cf40d8e65a42e90ec2a5ff90814dea936a24ccf9...,Senior-Engineer - Global Engineering Labs,Tanggungjawab You are responsible for new equi...,Equipment Maintenance,tsc,Basic,0.8498,0.670282,1,equipment usage,Advanced,"The job requires overseeing installation, modi...","{\n ""skillTitle"": ""Equipment Maintenance"",\n ...",NaN
3,158086cf40d8e65a42e90ec2a5ff90814dea936a24ccf9...,Senior-Engineer - Global Engineering Labs,Tanggungjawab You are responsible for new equi...,Preventive Maintenance Management,tsc,Advanced,0.8576,0.669205,1,Defines and writes preventative maintenance sc...,Intermediate,The job requires defining and writing preventa...,"{\n ""skillTitle"": ""Preventive Maintenance Man...",NaN
4,158086cf40d8e65a42e90ec2a5ff90814dea936a24ccf9...,Senior-Engineer - Global Engineering Labs,Tanggungjawab You are responsible for new equi...,Corrective Maintenance Management,tsc,Advanced,0.8381,0.668778,1,full text,Intermediate,"The job requires overseeing installation, modi...","{\n ""skillTitle"": ""Corrective Maintenance Man...",NaN


In [197]:
# Filter out records that were previously processed
job_skill_gpt_evaluation = job_skill_gpt_evaluation[~(job_skill_gpt_evaluation["gpt_proficiencyLevel"].isna() & job_skill_gpt_evaluation["gpt_error"].isna())]
print(job_skill_gpt_evaluation.shape)

(6400, 14)


In [198]:
job_skill_gpt_evaluation_merged = pd.concat([job_skill_gpt_evaluation, job_skill_gpt_evaluation_remaining], ignore_index=True)
print(job_skill_gpt_evaluation_merged.shape)
print(job_skill_gpt_evaluation_merged['job_id'].nunique())

(7379, 14)
1633


In [199]:
job_skill_gpt_evaluation_merged.head(5)
print(job_skill_gpt_evaluation_merged.isnull().sum())

job_id                     0
jobtitle                   0
jobdescription             0
skillTitle                 0
skillSource                0
proficiencyLevel         165
proficiencyScore         498
confidenceScore            0
skillOccurrence            0
phrase                     0
gpt_proficiencyLevel     206
gpt_rationale            206
gpt_raw_response         206
gpt_error               7173
dtype: int64


In [200]:
job_skill_gpt_evaluation_merged = job_skill_gpt_evaluation_merged.sort_values(by=['job_id', 'jobtitle'], ascending=[True, True])
job_skill_gpt_evaluation_merged.head(5)

,job_id,jobtitle,jobdescription,skillTitle,skillSource,proficiencyLevel,proficiencyScore,confidenceScore,skillOccurrence,phrase,gpt_proficiencyLevel,gpt_rationale,gpt_raw_response,gpt_error
3049,00024e1cafc7c2a20a7eeeed9d8caadecad47c24d9614c...,"Executive (Credit Control), Business Office",Tanggungjawab What s on Offer Mission: Provide...,Communication,ccs,Basic,0.5265,0.623796,1,communication,Intermediate,The job requires attending and following up on...,"{\n ""skillTitle"": ""Communication"",\n ""profic...",NaN
3754,002617a9d01b85bc0320f87b742de208648c90e3109f4a...,Junior Data Management Specialist,Job description Join Amaris and embark on a ne...,Data and Information Management,tsc,Advanced,0.5263,0.718152,4,data management,Intermediate,The job requires the candidate to 'analyze req...,"{\n ""skillTitle"": ""Data and Information Manag...",NaN
3755,002617a9d01b85bc0320f87b742de208648c90e3109f4a...,Junior Data Management Specialist,Job description Join Amaris and embark on a ne...,Problem Solving,ccs,Advanced,0.6607,0.639494,1,problem - solving,Intermediate,The job requires analyzing requirements holist...,"{\n ""skillTitle"": ""Problem Solving"",\n ""prof...",NaN
3756,002617a9d01b85bc0320f87b742de208648c90e3109f4a...,Junior Data Management Specialist,Job description Join Amaris and embark on a ne...,Business Requirements Mapping,tsc,Advanced,0.8545,0.626309,3,userbusiness requirements analysis and enhance...,Intermediate,The job requires collaborating with stakeholde...,"{\n ""skillTitle"": ""Business Requirements Mapp...",NaN
3757,002617a9d01b85bc0320f87b742de208648c90e3109f4a...,Junior Data Management Specialist,Job description Join Amaris and embark on a ne...,Sales Delivery,tsc,Intermediate,0.7519,0.622938,1,delivery,Basic,The job description does not mention sales del...,"{\n ""skillTitle"": ""Sales Delivery"",\n ""profi...",NaN


In [208]:
# Drop those rows that proficiencyLevel, gpt_proficiencyLevel is null
job_skill_gpt_evaluation_merged = job_skill_gpt_evaluation_merged.dropna(subset=['proficiencyLevel','proficiencyScore','gpt_proficiencyLevel'])
print(job_skill_gpt_evaluation_merged.shape)
print(job_skill_gpt_evaluation_merged['job_id'].nunique())
print(job_skill_gpt_evaluation_merged.isnull().sum())

(6687, 14)
1548
job_id                     0
jobtitle                   0
jobdescription             0
skillTitle                 0
skillSource                0
proficiencyLevel           0
proficiencyScore           0
confidenceScore            0
skillOccurrence            0
phrase                     0
gpt_proficiencyLevel       0
gpt_rationale              0
gpt_raw_response           0
gpt_error               6687
dtype: int64


In [202]:
# GPT evaluated proficiency breakdown
print(job_skill_gpt_evaluation_merged['gpt_proficiencyLevel'].value_counts())
print(job_skill_gpt_evaluation_merged['gpt_proficiencyLevel'].value_counts(normalize=True))

# PL proficiency extractor breakdown
print(job_skill_gpt_evaluation_merged['proficiencyLevel'].value_counts())
print(job_skill_gpt_evaluation_merged['proficiencyLevel'].value_counts(normalize=True))


gpt_proficiencyLevel
Intermediate    3076
Advanced        2223
Basic           1388
Name: count, dtype: int64
gpt_proficiencyLevel
Intermediate    0.459997
Advanced        0.332436
Basic           0.207567
Name: proportion, dtype: float64
proficiencyLevel
Intermediate    2506
Basic           2115
Advanced        2066
Name: count, dtype: int64
proficiencyLevel
Intermediate    0.374757
Basic           0.316285
Advanced        0.308958
Name: proportion, dtype: float64


In [203]:
# Performance evaluation starts here
from sklearn.metrics import f1_score, precision_score, classification_report

y_pred = np.array(job_skill_gpt_evaluation_merged['proficiencyLevel'])
y_true = np.array(job_skill_gpt_evaluation_merged['gpt_proficiencyLevel'])

# Performing encoding of labels!
label_map = {'Basic': 0, 'Intermediate': 1, 'Advanced': 2}
y_pred = np.array([label_map[l] for l in y_pred])
y_true = np.array([label_map[l] for l in y_true])

print(y_pred[:20])
print(y_true[:20])

[0 2 2 2 1 2 1 1 1 2 1 1 2 0 2 1 2 1 1 2]
[1 1 1 1 0 1 1 1 1 1 1 1 2 1 2 1 2 1 0 2]


In [204]:
f1 = f1_score(y_true, y_pred, average='weighted', zero_division=0)
precision = precision_score(y_true, y_pred, average='weighted', zero_division=0)

print(f'weighted f1 score is {f1}')
print(f'weighted precision score is {precision}')

weighted f1 score is 0.4083730302754131
weighted precision score is 0.42446752496711165


In [205]:
y_pred = np.array(job_skill_gpt_evaluation_merged['proficiencyLevel'])
y_true = np.array(job_skill_gpt_evaluation_merged['gpt_proficiencyLevel'])

print(classification_report(y_true, y_pred, target_names=['Basic', 'Intermediate', 'Advanced'], labels=['Basic', 'Intermediate', 'Advanced']))


              precision    recall  f1-score   support

       Basic       0.27      0.42      0.33      1388
Intermediate       0.48      0.39      0.43      3076
    Advanced       0.44      0.41      0.43      2223

    accuracy                           0.40      6687
   macro avg       0.40      0.41      0.40      6687
weighted avg       0.42      0.40      0.41      6687



Now i will need to do it for the slightly modified baseline algo..

In [215]:
job_skill_gpt_evaluation_modified = pd.read_csv('./data/jobpostings_skills_with_proficiency_labels_modified_algo.csv')
print(job_skill_gpt_evaluation_modified.shape)
print(job_skill_gpt_evaluation_modified['job_id'].nunique())

(7379, 14)
1633


In [216]:
print(job_skill_gpt_evaluation_modified.isnull().sum())

job_id                     0
jobtitle                   0
jobdescription             0
skillTitle                 0
skillSource                0
proficiencyLevel         165
proficiencyScore         498
confidenceScore            0
skillOccurrence            0
phrase                     0
gpt_proficiencyLevel       0
gpt_rationale              0
gpt_raw_response           0
gpt_error               7379
dtype: int64


In [217]:
# Drop those rows that proficiencyLevel, gpt_proficiencyLevel is null
job_skill_gpt_evaluation_modified = job_skill_gpt_evaluation_modified.dropna(subset=['proficiencyLevel','proficiencyScore'])
print(job_skill_gpt_evaluation_modified.shape)
print(job_skill_gpt_evaluation_modified['job_id'].nunique())
print(job_skill_gpt_evaluation_modified.isnull().sum())

(6881, 14)
1596
job_id                     0
jobtitle                   0
jobdescription             0
skillTitle                 0
skillSource                0
proficiencyLevel           0
proficiencyScore           0
confidenceScore            0
skillOccurrence            0
phrase                     0
gpt_proficiencyLevel       0
gpt_rationale              0
gpt_raw_response           0
gpt_error               6881
dtype: int64


In [218]:
# GPT evaluated proficiency breakdown
print(job_skill_gpt_evaluation_modified['gpt_proficiencyLevel'].value_counts())
print(job_skill_gpt_evaluation_modified['gpt_proficiencyLevel'].value_counts(normalize=True))

# PL proficiency extractor breakdown
print(job_skill_gpt_evaluation_modified['proficiencyLevel'].value_counts())
print(job_skill_gpt_evaluation_modified['proficiencyLevel'].value_counts(normalize=True))

gpt_proficiencyLevel
Intermediate    3151
Advanced        2298
Basic           1432
Name: count, dtype: int64
gpt_proficiencyLevel
Intermediate    0.457928
Advanced        0.333963
Basic           0.208109
Name: proportion, dtype: float64
proficiencyLevel
Intermediate    2533
Advanced        2272
Basic           2076
Name: count, dtype: int64
proficiencyLevel
Intermediate    0.368115
Advanced        0.330185
Basic           0.301700
Name: proportion, dtype: float64


In [219]:
# Performance evaluation starts here
from sklearn.metrics import f1_score, precision_score, classification_report

y_pred = np.array(job_skill_gpt_evaluation_modified['proficiencyLevel'])
y_true = np.array(job_skill_gpt_evaluation_modified['gpt_proficiencyLevel'])

# Performing encoding of labels!
label_map = {'Basic': 0, 'Intermediate': 1, 'Advanced': 2}
y_pred = np.array([label_map[l] for l in y_pred])
y_true = np.array([label_map[l] for l in y_true])

print(y_pred[:20])
print(y_true[:20])

[0 1 1 2 2 2 1 2 0 1 0 2 1 1 0 2 2 2 1 2]
[1 1 2 2 2 1 1 1 1 1 1 1 1 1 0 0 1 2 2 2]


In [220]:
f1 = f1_score(y_true, y_pred, average='weighted', zero_division=0)
precision = precision_score(y_true, y_pred, average='weighted', zero_division=0)

print(f'weighted f1 score is {f1}')
print(f'weighted precision score is {precision}')

weighted f1 score is 0.40343579730527523
weighted precision score is 0.41719641981535055
